ql/experimental/termstructures/crosscurrencyratehelpers.cpp

```diff
@@ -18,44 +18,154 @@
+            Schedule sch = legSchedule(evaluationDate, tenor, idx->tenor(), fixingDays, calendar,
+                                       convention, endOfMonth);
+            return IborLeg(sch, idx).withNotionals(1.0);
+        }
+
+        Real npvConstNotionalLeg(const Leg& iborLeg,
+                                 const Handle<YieldTermStructure>& discountCurveHandle) {
+            Date refDate = discountCurveHandle->referenceDate();
+            const YieldTermStructure& discountRef = **discountCurveHandle;
+            bool includeSettleDateFlows = true;
+            return CashFlows::npv(iborLeg, discountRef, includeSettleDateFlows, refDate) +
+                   discountRef.discount(iborLeg.back()->date()) - 1.0;
+        }
+
+        Real bpsConstNotionalLeg(const Leg& iborLeg,
+                                 const Handle<YieldTermStructure>& discountCurveHandle) {
+            const Spread basisPoint = 1.0e-4;
+            Date refDate = discountCurveHandle->referenceDate();
+            const YieldTermStructure& discountRef = **discountCurveHandle;
+            bool includeSettleDateFlows = true;
+            return CashFlows::bps(iborLeg, discountRef, includeSettleDateFlows, refDate) /
+                   basisPoint;
+        }
+
+        class ResettingLegHelper {
+          public:
+            explicit ResettingLegHelper(const YieldTermStructure& discountCurve,
+                                        const YieldTermStructure& foreignCurve)
+            : discountCurve_(discountCurve), foreignCurve_(foreignCurve) {}
+            Real discount(const Date& d) const {
+                return discountCurve_.discount(d);
+            }
+            Real adjustedNotional(const Date& d) const {
+                return foreignCurve_.discount(d) / discountCurve_.discount(d);
+            }
+
+          private:
+            const YieldTermStructure& discountCurve_;
+            const YieldTermStructure& foreignCurve_;
+        };
+
+        class ResettingLegNPVCalculator : public AcyclicVisitor, public Visitor<Coupon> {
+          public:
+            explicit ResettingLegNPVCalculator(const YieldTermStructure& discountCurve,
+                                               const YieldTermStructure& foreignCurve)
+            : helper_(discountCurve, foreignCurve), npv_(0.0) {}
+            void visit(Coupon& c) override {
+                Date start = c.accrualStartDate();
+                Date end = c.accrualEndDate();
+                Time accrual = c.accrualPeriod();
+                Real npv =
+                    helper_.adjustedNotional(start) *
+                    (-helper_.discount(start) + helper_.discount(end) * (1.0 + c.rate() * accrual));
+                npv_ += npv;
+            }
+            Real NPV() const { return npv_; }
+
+          private:
+            ResettingLegHelper helper_;
+            Real npv_;
+        };
+
+        class ResettingLegBPSCalculator : public AcyclicVisitor, public Visitor<Coupon> {
```

Comment:
The `ResettingLegNPVCalculator` can calculate the BPS in the same loop as the NPV; it has all the information already so the cost would be minimal (as opposed to doing another loop)

Reply:
I struggled a bit to incorporate this suggestion. While the NPV of the resetting leg is always needed in the pricing formula, the BPS of that leg is not always required. That depends on the type of leg with the spread, which can also be present in the constant notional leg. So, I thought it makes sense to keep the two separate (just like with the constant notional leg)...
On the other hand, we could have the `ResettingLegNPVCalculator` calculate the BPS simultaneously with the NPV (as suggested) and return them both by reference in a `npvbpsResettingLeg` kind of function similar to how `CashFlows::npvbps` does it. For the sake of consistency same could be done with the constant notional leg. Then, in the `impliedQuote` method BPS for both quote and base currency will always be calculated and only selected depending on where the spread is. I can try that.

Reply:
Ok, the changes are in. I hope they are in line with your suggestion. Thanks.

Reply:
Yes, thanks.  It's true that one of the BPS is unused, but calculating it together with the NPV has practically no cost, and we get to reduce the number of helper classes.